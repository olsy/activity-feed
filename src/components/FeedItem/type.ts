export enum Type {
  Meeting = 'MeetingNote',
  Call = 'Phone',
  Beer = 'Beer',
  Coffee = 'Coffee',
  Note = 'Message',
}
