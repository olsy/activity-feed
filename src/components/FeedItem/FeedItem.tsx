import { Box, Stack, Typography } from '@mui/material';
import { styled } from '@mui/material/styles';
import Icon from './components/Icon';
import Title from './components/Title';
import { Type } from './type';

const Wrapper = styled(Stack)(() => ({
  position: 'relative',
  '&::before': {
    content: '""',
    position: 'absolute',
    top: 0,
    bottom: 0,
    left: '54px',
    width: '2px',
    background: '#ECECEC',
  },
}));

const TextWrapper = styled(Box)(() => ({
  background: '#F6F6F6',
  padding: '12px 16px',
  borderRadius: '6px',
  flex: 1,
}));

type Props = {
  type: Type;
  from: string;
  to: string;
  message: string;
  timestamp: string;
};

const FeedItem = ({ type, from, to, message, timestamp }: Props) => {
  return (
    <Wrapper direction={'row'} alignItems="center" spacing={2} sx={{ padding: '6px 0' }}>
      <Box>
        <Stack direction={'row'} alignItems="center">
          <Typography mr={2}>{timestamp}</Typography>
          <Icon type={type} />
        </Stack>
      </Box>
      <TextWrapper>
        <Title type={type} from={from} to={to} />
        <Typography sx={{ fontSize: '0.8em' }}>{message}</Typography>
      </TextWrapper>
    </Wrapper>
  );
};

export default FeedItem;
