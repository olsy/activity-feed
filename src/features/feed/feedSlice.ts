import type { PayloadAction } from '@reduxjs/toolkit';
import { createSelector, createSlice } from '@reduxjs/toolkit';
import { RootState } from '../../app/store';

export enum Type {
  Meeting = 'MeetingNote',
  Call = 'Phone',
  Beer = 'Beer',
  Coffee = 'Coffee',
  Note = 'Message',
  List = 'New',
}

export interface FeedItem {
  id: number;
  type: Type;
  from: string;
  to: string;
  message: string;
  timestamp: string;
}

export interface feedState {
  count: number;
  data: FeedItem[];
}

const mock = [
  {
    id: 1,
    type: Type.Meeting,
  },
  {
    id: 2,
    type: Type.Call,
  },
  {
    id: 3,
    type: Type.Coffee,
  },
  {
    id: 4,
    type: Type.Beer,
  },
  {
    id: 5,
    type: Type.Note,
  },
].map((el) => ({
  ...el,
  from: 'You',
  to: 'Milton Romaguera',
  timestamp: new Date().toISOString(),
  message: 'And a more formal meeting',
}));

const initialState: feedState = {
  count: mock.length || 0,
  data: mock || [],
};

export const feedSlice = createSlice({
  name: 'feed',
  initialState,
  reducers: {
    add: (state, action: PayloadAction<Omit<FeedItem, 'id'>>) => {
      state.count++;
      state.data.unshift({ ...action.payload, id: state.count });
    },
    removeById: (state, action: PayloadAction<number>) => {
      const idx = state.data.findIndex((el) => el.id === action.payload);
      state.data.splice(idx, 1);
    },
  },
});

// Action creators are generated for each case reducer function
export const { removeById, add } = feedSlice.actions;

export default feedSlice.reducer;

export const baseFeedSelector = (state: RootState) => state.feed;
export const feedSelector = createSelector([baseFeedSelector], (feed): FeedItem[] => feed.data);
