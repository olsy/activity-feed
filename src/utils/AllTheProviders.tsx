import { CssBaseline, ThemeProvider } from '@mui/material';
import React from 'react';
import { Provider } from 'react-redux';
import { store } from '../app/store';
import theme from '../themes/default';

const AllTheProviders = ({ children }: { children: React.ReactNode }) => (
  <Provider store={store}>
    <ThemeProvider theme={theme}>
      {/* CssBaseline kickstart an elegant, consistent, and simple baseline to build upon. */}
      <CssBaseline />
      {children}
    </ThemeProvider>
  </Provider>
);

export default AllTheProviders;
