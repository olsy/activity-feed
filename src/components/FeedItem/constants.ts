import { Type } from '../../features/feed/feedSlice';

export const TitleParts = {
  [Type.Meeting]: ` had a meeting with `,
  [Type.Call]: ` had a call with `,
  [Type.Beer]: ` had a beer with `,
  [Type.Coffee]: ` had a coffee with `,
  [Type.Note]: ` had a note with `,
};
