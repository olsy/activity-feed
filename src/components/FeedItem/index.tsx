import { Box, Typography } from '@mui/material';
import { useDispatch } from 'react-redux';
import { removeById, Type } from '../../features/feed/feedSlice';
import FeedContainer from '../FeedContainer';
import Title from './components/Title';

interface Props {
  id: number;
  type: Type;
  from: string;
  to: string;
  message: string;
  timestamp: string;
}

const FeedItem = ({ id, type, from, to, message, timestamp }: Props) => {
  const dispatch = useDispatch();

  return (
    <FeedContainer type={type} timestamp={timestamp} remove={() => dispatch(removeById(id))}>
      <Box>
        <Title type={type} from={from} to={to} />
        <Typography sx={{ fontSize: '0.8em' }} data-testid="message">
          {message}
        </Typography>
      </Box>
    </FeedContainer>
  );
};

export default FeedItem;
