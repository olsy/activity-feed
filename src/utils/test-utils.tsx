import { render, RenderOptions } from '@testing-library/react';
import { ReactElement } from 'react';
import AllTheProviders from './AllTheProviders';

export const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';
