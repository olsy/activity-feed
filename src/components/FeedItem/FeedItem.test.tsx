import { screen } from '@testing-library/react';
import { Type } from '../../features/feed/feedSlice';
import { customRender } from '../../utils/test-utils';
import { TitleParts } from './constants';
import FeedItem from './index';

const props = {
  id: 1,
  type: Type.Call,
  from: 'Me',
  to: 'You',
  message: 'hello world',
  timestamp: new Date().toISOString(),
};

describe('<FeedItem />', () => {
  it('should match snapshot', () => {
    expect(customRender(<FeedItem {...props} />)).toMatchSnapshot();
  });

  beforeEach(() => {
    customRender(<FeedItem {...props} />);
  });

  it('renders component', () => {
    const title = screen.getByTestId('title');

    expect(title).toBeInTheDocument();

    expect(title?.textContent).toEqual(`${props.from}${TitleParts[props.type]}${props.to}`);

    const message = screen.getByTestId('message');

    expect(message).toBeInTheDocument();

    expect(message?.textContent).toEqual(props.message);

    const timestamp = screen.getByTestId('timestamp');

    expect(timestamp).toBeInTheDocument();

    expect(timestamp?.textContent).toEqual('a few seconds');
  });
});
