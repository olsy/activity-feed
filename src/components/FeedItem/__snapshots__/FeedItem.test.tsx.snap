// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FeedItem /> should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiStack-root css-9c4v0-MuiStack-root"
      >
        <div
          class="MuiBox-root css-0"
        >
          <div
            class="MuiStack-root css-1dksopf-MuiStack-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-1t382tg-MuiTypography-root"
              data-testid="timestamp"
            >
              a few seconds
            </p>
            <div
              class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-v6eczs-MuiAvatar-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="PhoneIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6.62 10.79c1.44 2.83 3.76 5.14 6.59 6.59l2.2-2.2c.27-.27.67-.36 1.02-.24 1.12.37 2.33.57 3.57.57.55 0 1 .45 1 1V20c0 .55-.45 1-1 1-9.39 0-17-7.61-17-17 0-.55.45-1 1-1h3.5c.55 0 1 .45 1 1 0 1.25.2 2.45.57 3.57.11.35.03.74-.25 1.02z"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root css-adclsq"
        >
          <div
            class="MuiBox-root css-0"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-t60ksl-MuiTypography-root"
              data-testid="title"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 css-tzy8sz-MuiTypography-root"
              >
                Me
              </span>
               had a call with 
              <span
                class="MuiTypography-root MuiTypography-body1 css-tzy8sz-MuiTypography-root"
              >
                You
              </span>
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1 css-16v3jx0-MuiTypography-root"
              data-testid="message"
            >
              hello world
            </p>
          </div>
          <div
            class="css-1opzxad"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-bfk5jd-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="EditIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-bfk5jd-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="DeleteIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="MuiStack-root css-9c4v0-MuiStack-root"
      >
        <div
          class="MuiBox-root css-0"
        >
          <div
            class="MuiStack-root css-1dksopf-MuiStack-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-1t382tg-MuiTypography-root"
              data-testid="timestamp"
            >
              a few seconds
            </p>
            <div
              class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-v6eczs-MuiAvatar-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="PhoneIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6.62 10.79c1.44 2.83 3.76 5.14 6.59 6.59l2.2-2.2c.27-.27.67-.36 1.02-.24 1.12.37 2.33.57 3.57.57.55 0 1 .45 1 1V20c0 .55-.45 1-1 1-9.39 0-17-7.61-17-17 0-.55.45-1 1-1h3.5c.55 0 1 .45 1 1 0 1.25.2 2.45.57 3.57.11.35.03.74-.25 1.02z"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root css-adclsq"
        >
          <div
            class="MuiBox-root css-0"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-t60ksl-MuiTypography-root"
              data-testid="title"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 css-tzy8sz-MuiTypography-root"
              >
                Me
              </span>
               had a call with 
              <span
                class="MuiTypography-root MuiTypography-body1 css-tzy8sz-MuiTypography-root"
              >
                You
              </span>
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1 css-16v3jx0-MuiTypography-root"
              data-testid="message"
            >
              hello world
            </p>
          </div>
          <div
            class="css-1opzxad"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-bfk5jd-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="EditIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-bfk5jd-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="DeleteIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiStack-root css-9c4v0-MuiStack-root"
    >
      <div
        class="MuiBox-root css-0"
      >
        <div
          class="MuiStack-root css-1dksopf-MuiStack-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-1t382tg-MuiTypography-root"
            data-testid="timestamp"
          >
            a few seconds
          </p>
          <div
            class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-v6eczs-MuiAvatar-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="PhoneIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M6.62 10.79c1.44 2.83 3.76 5.14 6.59 6.59l2.2-2.2c.27-.27.67-.36 1.02-.24 1.12.37 2.33.57 3.57.57.55 0 1 .45 1 1V20c0 .55-.45 1-1 1-9.39 0-17-7.61-17-17 0-.55.45-1 1-1h3.5c.55 0 1 .45 1 1 0 1.25.2 2.45.57 3.57.11.35.03.74-.25 1.02z"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root css-adclsq"
      >
        <div
          class="MuiBox-root css-0"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-t60ksl-MuiTypography-root"
            data-testid="title"
          >
            <span
              class="MuiTypography-root MuiTypography-body1 css-tzy8sz-MuiTypography-root"
            >
              Me
            </span>
             had a call with 
            <span
              class="MuiTypography-root MuiTypography-body1 css-tzy8sz-MuiTypography-root"
            >
              You
            </span>
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-16v3jx0-MuiTypography-root"
            data-testid="message"
          >
            hello world
          </p>
        </div>
        <div
          class="css-1opzxad"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-bfk5jd-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="EditIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-bfk5jd-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="DeleteIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
